Extension/src/LanguageServer/client.ts

```diff
@@ -616,12 +622,31 @@ class DefaultClient implements Client {
         for (let e of editors) {
             e.visibleRanges.forEach(range => ranges.push(Range.create(range.start.line, range.start.character, range.end.line, range.end.character)));
         }
-        let params: DidChangeVisibleRangesParams = {
-            uri: uri.toString(),
-            ranges: ranges
-        };
-        this.notifyWhenReady(() => this.languageClient.sendNotification(DidChangeVisibleRangesNotification, params));
-}
+
+        // Only send ranges if they have actually changed.
+        let isSame: boolean = false;
+        let savedRanges: Range[] = this.visibleRanges.get(uri.toString());
+        if (savedRanges) {
+            if (ranges.length === savedRanges.length) {
+                for (let i: number = 0; i < ranges.length; i++) {
+                    if (ranges[i] !== savedRanges[i]) {
+                        break;
+                    }
+                }
+                isSame = true;
+            }
+        } else {
+            isSame = ranges.length === 0;
+        }
+        if (!isSame) {
+            this.visibleRanges.set(uri.toString(), ranges);
+            let params: DidChangeVisibleRangesParams = {
+                uri: uri.toString(),
+                ranges: ranges
+            };
+            this.notifyWhenReady(() => this.languageClient.sendNotification(DidChangeVisibleRangesNotification, params));
```

Comment:
This doesn't work still. When I put a breakpoint here it starts working though, so there's some sort of race condition.

Reply:
I'm not able to repro an issue here, and don't see where there could be a race.  When you get a chance, I'd like to see your repro.